packages/statemanager/src/ethersStateManager.ts

```diff
@@ -0,0 +1,391 @@
+  private stateRoot: Buffer | undefined
+  _cache: Cache
+
+  constructor(opts: EthersStateManagerOpts) {
+    super({})
+    if (typeof opts.provider === 'string') {
+      this.provider = new StaticJsonRpcProvider(opts.provider)
+    } else if (opts.provider instanceof JsonRpcProvider) {
+      this.provider = opts.provider
+    } else {
+      throw new Error(`valid JsonRpcProvider or url required; got ${opts.provider}`)
+    }
+
+    this.blockTag =
+      typeof opts.blockTag === 'bigint' ? bigIntToHex(opts.blockTag) : opts.blockTag ?? 'latest'
+
+    this.contractCache = new Map()
+    this.storageCache = new Map()
+
+    /*
+     * For a custom StateManager implementation adopt these
+     * callbacks passed to the `Cache` instantiated to perform
+     * the `get`, `put` and `delete` operations with the
+     * desired backend.
+     */
+    const getCb: getCb = async (address) => {
+      return this.getAccountFromProvider(address)
+    }
+    const putCb: putCb = async (_keyBuf, _accountRlp) => {
+      return Promise.resolve()
+    }
+    const deleteCb = async (_keyBuf: Buffer) => {
+      return Promise.resolve()
+    }
+    this._cache = new Cache({ getCb, putCb, deleteCb })
+  }
+
+  copy(): EthersStateManager {
+    const newState = new EthersStateManager({
+      provider: this.provider,
+      blockTag: isHexPrefixed(this.blockTag)
+        ? BigInt(this.blockTag)
+        : (this.blockTag as BlockTagType),
+    })
+    ;(newState as any).contractCache = new Map(this.contractCache)
+    ;(newState as any).storageCache = new Map(this.storageCache)
+    ;(newState as any)._cache = this._cache
+    return newState
+  }
+
+  /**
+   * Sets the new block tag used when querying the provider and clears the
+   * internal cache.
+   * @param blockTag - the new block tag to use when querying the provider
+   */
+  setBlockTag(blockTag: BlockTagType): void {
+    if (typeof blockTag === 'bigint') {
+      this.blockTag = bigIntToHex(blockTag)
+    } else {
+      this.blockTag = blockTag
+    }
+    this.clearCache()
+  }
+
+  /**
+   * Clears the internal cache so all accounts, contract code, and storage slots will
+   * initially be retrieved from the provider
+   */
+  clearCache(): void {
+    this.contractCache.clear()
+    this.storageCache.clear()
+    this._cache.clear()
+  }
+
+  /**
+   * Gets the code corresponding to the provided `address`.
+   * @param address - Address to get the `code` for
+   * @returns {Promise<Buffer>} - Resolves with the code corresponding to the provided address.
+   * Returns an empty `Buffer` if the account has no associated code.
+   */
+  async getContractCode(address: Address): Promise<Buffer> {
+    let codeBuffer = this.contractCache.get(address.toString())
+    if (codeBuffer !== undefined) return codeBuffer
+    const code = await this.provider.getCode(address.toString(), this.blockTag)
+    codeBuffer = toBuffer(code)
+    this.contractCache.set(address.toString(), codeBuffer)
```

Comment:
```suggestion
    this.putContractCode(address, codeBuffer)
```
nbd, but this aligns better with what we're doing in getContractStorage